src/buffer_par.rs

```diff
@@ -0,0 +1,523 @@
+use rayon::iter::plumbing::*;
+use rayon::iter::{IndexedParallelIterator, ParallelIterator};
+use rayon::slice::{ChunksExact, ChunksExactMut, ParallelSlice, ParallelSliceMut};
+use std::fmt;
+use std::ops::{Deref, DerefMut};
+
+use crate::traits::Pixel;
+use crate::ImageBuffer;
+
+/// Parallel iterator over pixel refs.
+pub struct PixelsPar<'a, P>
+where
+    P: Pixel + Sync + 'a,
+    P::Subpixel: Sync + 'a,
+{
+    chunks: ChunksExact<'a, P::Subpixel>,
+}
+
+impl<'a, P> ParallelIterator for PixelsPar<'a, P>
+where
+    P: Pixel + Sync + 'a,
+    P::Subpixel: Sync + 'a,
+{
+    type Item = &'a P;
+
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+    where
+        C: UnindexedConsumer<Self::Item>,
+    {
+        self.chunks
+            .map(|v| <P as Pixel>::from_slice(v))
+            .drive_unindexed(consumer)
+    }
+
+    fn opt_len(&self) -> Option<usize> {
+        Some(self.len())
+    }
+}
+
+impl<'a, P> IndexedParallelIterator for PixelsPar<'a, P>
+where
+    P: Pixel + Sync + 'a,
+    P::Subpixel: Sync + 'a,
+{
+    fn drive<C: Consumer<Self::Item>>(self, consumer: C) -> C::Result {
+        self.chunks
+            .map(|v| <P as Pixel>::from_slice(v))
+            .drive(consumer)
+    }
+
+    fn len(&self) -> usize {
+        self.chunks.len()
+    }
+
+    fn with_producer<CB: ProducerCallback<Self::Item>>(self, callback: CB) -> CB::Output {
+        self.chunks
+            .map(|v| <P as Pixel>::from_slice(v))
+            .with_producer(callback)
+    }
+}
+
+impl<P> Clone for PixelsPar<'_, P>
+where
+    P: Pixel + Sync,
+    P::Subpixel: Sync,
+{
+    fn clone(&self) -> Self {
+        PixelsPar {
+            chunks: self.chunks.clone(),
+        }
+    }
+}
+
+impl<P> fmt::Debug for PixelsPar<'_, P>
```

Comment:
This also might be automatically derivable?

Reply:
Just like clone, `derive(Debug)` also seems to put unnecessary bounds on `P` (when it should be just `P::Subpixel`, I think), and also breaks on `EnumeratePixelsPar` and `EnumeratePixelsMutPar` for some reason.